docs/DesignDocs/SGX_0-base_Enclave.md

```diff
@@ -0,0 +1,228 @@
+much higher than the typical threshold, especially for Linux, makes sure the
+enclave can be loaded on most systems.  The Enclave Loader does not need to
+reserve the linear address range between `SECS.BASEADDR`=0 and `SECS.SIZE` for
+the Enclave. It should reserve the memory range starting from `ENCLAVE_START`,
+in this case, predefined at Enclave signing time, with a size smaller than
+`SECS.SIZE`.
+
+It's possible to support multiple 0-base Enclaves in a single process. The SGX
+CPU does not require each Enclave, even in a single process, to have an unique
+Enclave Linear Range, not overlapping with any other Enclaves. The SGX CPU does
+enforce that any EPC page must only belong to a single Enclave, the code inside
+an Enclave can only access EPC pages belong to itself, and any linear address to
+EPC page mapping cannot be changed after the EPC page is added to an Enclave.
+With these policies enforced by the SGX CPU, if a malicious OS  manipulates the
+Page Table Entries and attempts to circumvent Enclave memory access control or
+to trick the code inside an Enclave to access EPC pages belong to another
+Enclave, the attack will be detected and blocked. Within any committed page in a
+process, a normal-behaving OS only allows a single mapping (between the linear
+address and the physical address) for any specific linear address. As a result,
+when running multiple Enclaves within a single process, each Enclave must have a
+unique, non-overlapping [`ENCLAVE_START`, `ENCLAVE_END`] range (linear address
+of each committed EPC pages of an Enclave is within that range), within the
+process. For a 0-base Enclave, The linear address range between
+`SECS.BASEADDR`=0 and `ENCLAVE_START` does not have any committed EPC page, nor
+is reserved by the Enclave Loader. From the OS' point of view, that linear
+address range is available for another 0-base Enclave's reserved memory range,
+as long as the second enclave's [`ENCLAVE_START`, `ENCLAVE_END`] range can fit
+within.  For each 0-base Enclave within a process, the ECALL/OCALL marshalling
+buffer and any shared memory an Enclave needs to access must be allocated above
+the Enclave's `SECS.SIZE` value.
+
+Running multiple instance of a single 0-base Enclave in a single process won't
+be possible, as each instance would has to be loaded at exactly the same
+[`ENCLAVE_START`, `ENCLAVE_END`] range to maintain the Enclave measurement,
+which is not possible within a single process.
+
+## Support SGX Enclave with SECS.BASEADDR set to 0
+
+To support "0-base" SGX Enclave with `SECS.BASEADDR`=0, an flag indicating
+"0-base", and a variable for the predetermined `ENCLAVE_START` value are added
+to `oe_sgx_enclave_config` and `_oesgx_load_context`.
+
+```C
+typedef struct oe_sgx_enclave_config_t
+{
+    uint16_t product_id;
+    uint16_t security_version;
+
+    /* Flag to indicate 0-base enclave, keeping alignment */
+    uint32_t flag_0_base;
```

Comment:
Follow up to SIG-arch discussion: given that `start_addr` must never be zero when `flag_0_base` is set, is there any scenario where we want to support a fixed `start_addr` without 0-base as measured in the properties?

If not, can this be inferred from `start_addr != 0`?

Reply:
A predetermined `start_addr` might have value outside of the "0-base" Enclave context, for example, to support loading legacy non-position-independent apps in the Enclave (libOS model) at specific linear address. A **preferred** `start_addr` might also make sense. I think having a flag and a `start_addr` allows more flexibility in the future.  The comments for `start_addr` were changed in the updated PR.